### DISCUSSION TITLE
Extend arrays with a remove functionality.

### DISCUSSION THREAD

**Discussion by MaxTymchii** [jingoro2112/wrench#26] [Ideas]
Hi there,

As I can see `wrench` does not support such an important action as removing some items from an array as hash-table for example. 
I tried to set `null` for the corresponding item but `null` is `0` so it's not what I want as a result.

Now, I create a new array without this item by copying all the elements, but it's definitely how it should work :)

**Comment by jingoro2112** [comment]
This already exists, the "._remove" function:

I consider this a failure of the documentation :( apologies for not making
it more clear.

*compiler-intrinsics* Some compiler directives are included for working
with arrays and hash tables:

hashTable = { 1:"one", 2:"two", 3:3, "str":6 };

//  ._count
print( hashTable._count ); // prints 4

// ._exists
hashTable._exists( 2 ); // returns 'true'
hashTable._exists( 20 ); // returns 'false

// ._remove
hashTable._exists( 2 ); // returns 'true'
hashTable._remove( 2 );
hashTable._exists( 2 ); // now false

**Reply by MaxTymchii** [reply]
I know about a hash table and it works for me. I am talking now about an array. after tiny test I have also a question for hash table actually too.


```
var hashTable = {:};
hashTable["mock"] = "value";
hashTable["mock2"] = "value2";
hashTable["mock3"] = "value3";

var array = {1, 2, 3, 4, 5};


hashTable._remove("mock");
array._remove(2);


println( "hashTable" );
for ( var v, var k : hashTable ) {
    print(v);
    println(k);
}

println( "array" );
for ( var v : array ) {
    println(v);
}

```

Result:

```
hashTable
00
0value3
0value2
array
1
2
3
4
5
```


As you can see 2 value on array is available.  Hashtable shows really interesting behavior.


P.S.: I think I can try to help you with examples on your website if you want.

Regards,
Max.

**Comment by jingoro2112** [comment]
OH ... removing from arrays? well.. so you mean turning:
[ 0, 1, 2, 3 ]
into
[ 0, 2, 3 ]
by removing element 1?

I could have sworn I added that.. yeah oops..

it's part of the container lib that I stubbed out and never finished!!

array::count()
array::remove( array, where [,count] )
array::insert( array, where [,count] )
hash::exists( hash, key )
hash::remove( hash, key )

Guess it's time for me to finish it :P I'll make it part of my next update,
I'm halfway through the parser re-work, so declarations like this will work:

var a = { 1, 2, { 3.1, 3.2, 3.3 }, { 4:"four"} };

nested container declaration :)

If I'm lucky I'll have it done and unit-tested within a day or so!

-Curt

**Reply by MaxTymchii** [reply]
Please, also check my print logs of a hashtable, as you can see it's an extraordinary thing with keys after removing one of them that has an almost identical name!

**Comment by jingoro2112** [comment]
it would appear foreach key and value is broken! I will fix this
immediately and put in unit tests, also the array remove is not working
either.. :(

**Comment by jingoro2112** [comment]
Okay should be fixed in the latest 5.0.1 push:

var hashTable = {:};
hashTable["mock"] = "value";
hashTable["mock2"] = "value2";
hashTable["mock3"] = "value3";

var array = {1, 2, 3, 4, 5};


hashTable._remove("mock");
array._remove(2);


println( "hashTable" );
for ( var v, var k : hashTable ) {
    print(v);
    println(k);
}

println( "array" );
for ( var v : array ) {
    println(v);
}

Result:


hashTable
mock3value3
mock2value2
array
1
2
4
5

**Reply by ander-theo** [reply]
Note to update an error in the documentation, in section foreach:
```
for( var k,v : somArray )
{
   // same as above but k will take on the value 0, 1 and 2
}
```
(https://home.workshopfriends.com/wrench/www/#:~:text=for(%20var%20k%2Cv%20%3A%20somArray%20)%0A%7B%0A%20%20%20//%20same%20as%20above%20but%20k%20will%20take%20on%20the%20value%200%2C%201%20and%202%0A%7D)

I was struggling with this failing until I saw what you wrote above:
for ( var v, **_var_** k : hashTable ) {
    print(v);
    println(k);
}

**Comment by MaxTymchii** [comment]
Big-big thanks to you. Now everything works perfectly!

**Comment by jingoro2112** [comment]
thank you!
